### ISSUE TITLE
validation does not test all API requirements

### ISSUE THREAD

**Issue by christopher-johnson** [IIIF/presentation-validator#56]
In the Presentation 2.1 API spec, there are 157 "MUSTS".  I think that a proper API validation tool should have a test for each MUST.   For example, "A reference to the Image API context document must be included".   WIth this [method](https://github.com/sul-dlss/iiif-evented-canvas/blob/master/src/imageResourceFactory.js#L36-L45), this is now expected in Mirador 2.5.1.

Yet this tool does not test for this and "validates" manifests that are not valid.

I believe that a real IIIF manifest test suite is needed, so that implementers (and client developers) can actually validate manifests against the current API specification.  I would suggest using [EARL](https://www.w3.org/TR/EARL10-Guide/) and modelling it after the[ jsonld test suite](https://json-ld.org/test-suite/).

**Comment by christopher-johnson** [comment]
[SHACL](https://www.w3.org/TR/shacl/) could also be used.  Here is an example that tests the Image API `exif:width` and `exif:height` properties for cardinality, datatype and inclusion.

```turtle
@prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .
@prefix rdfs: <http://www.w3.org/2000/01/rdf-schema#> .
@prefix sh: <http://www.w3.org/ns/shacl#> .
@prefix xsd: <http://www.w3.org/2001/XMLSchema#> .
@prefix ex: <http://example.com/ns#> .
@prefix iiif: <http://iiif.io/api/image/2#> .
@prefix exif: <http://www.w3.org/2003/12/exif/ns#> .

ex:ImageServiceShape
	a sh:NodeShape ;
        sh:targetClass iiif:ImageService ;
	sh:property [
		sh:path exif:width ;
		sh:maxCount 1 ;
        sh:or ( [ sh:datatype xsd:string ] [ sh:datatype xsd:integer ] ) ;
        sh:minInclusive 0 ;
        sh:maxInclusive 99999 ;
	] ;
	sh:property [
		sh:path exif:height ;
		sh:maxCount 1 ;
        sh:or ( [ sh:datatype xsd:string ] [ sh:datatype xsd:integer ] ) ;
        sh:minInclusive 0 ;
        sh:maxInclusive 99999 ;
	] .`
```
with this input:
```jsonld
{
  "@context" : "http://iiif.io/api/image/2/context.json",
  "@id" : "http://www.example.org/image-service/abcd1234",
  "@type": "iiif:ImageService",
  "width" : "300000",
  "height" : 4000
}
```
produces this report:
```turtle
[
	a sh:ValidationResult ;
	sh:resultSeverity sh:Violation ;
	sh:sourceConstraintComponent sh:MaxInclusiveConstraintComponent ;
	sh:sourceShape _:n125 ;
	sh:focusNode <http://www.example.org/image-service/abcd1234> ;
	sh:value 300000 ;
	sh:resultPath exif:width ;
	sh:resultMessage "Value is not <= 99999" ;
] .
```
Using EARL and SHACL for API validation would provide a nice W3C-based specification compliance methodology for the IIIF.  

I work on this independently.  At UB Leipzig, diagnosing and reporting IIIF implementation issues scientifically is a primary task at the moment.

**Comment by azaroth42** [comment]
What test would find the missing `@context` URI, that in turn would not incorrectly penalize a non IIIF service associated with an image resource?

(For reference, the requirement is here: http://iiif.io/api/presentation/2.1/#image-resources )

**Comment by zimeon** [comment]
I think SHACL tests would be good, but as @azaroth42 notes, there are additional JSON-specific (rather than RDF) requirements that would have to be done is some other way (such as at present).

Aside - there are valid images out there with size >= 100,000 pixels! e.g. https://libimages.princeton.edu/loris/pudl0052/6131707/00000001.jp2/info.json

**Comment by christopher-johnson** [comment]
@azaroth42 I believe that the Presentation API should provide `@type` for the service class.  Similarly, an Image API service does not provide an `@type`, though doing so would make the class relationship reasonable ontologically.  Would not it be clearer to scope the service context with `@type`?  See [PR #462 ](https://github.com/json-ld/json-ld.org/pull/462)  and [example 54](https://json-ld.org/spec/latest/json-ld/#scoped-contexts)

Furthermore, as I indicated sometime ago, embedded `@context` cannot be generated fromRDF as it is not a property URI, even though it is legal JSON-LD.

**Comment by christopher-johnson** [comment]
@zimeon true, but the server throws `Server Side Error: encoder error -2 when writing image file `  when it is retrieved (for me)...

I agree that the service should be able to make it is own constraints on size, however.  The SHACL example just illustrates that a JSON string can be validated numerically (the presumption of a max size was arbitrary).

**Comment by azaroth42** [comment]
@christopher-johnson But the spec doesn't say that. So the validator is doing the best as it can, according to the 2.x specifications.  In 3.x I agree that there is a strong argument for making `type` REQUIRED on all resources.

**Comment by azaroth42** [comment]
A set of invalid manifests that are validated by the service would be very useful.  Or even just snippets that we can build into test cases.  Feel free to make one issue per incorrect validation.

**Comment by azaroth42** [comment]
To the SHACL suggestion ... the context is not semantically versioned in the same way as the API specifications are.  For example, we swapped `conformsTo` and `implements` here:  https://github.com/IIIF/iiif.io/commit/9f682681c22c405627dff2502da722f5bfd78e89#diff-f35d25db037fe4c3682c57762da34fef outside of a version update. That would then invalidate the SHACL tests. 

For 3.x, once we have a more constrained JSON representation, JSON-Schema might be more appropriate for a comprehensive set of tests.

**Comment by christopher-johnson** [comment]
As a point of reference for this issue, I have been working on a generalized [JSON-LD API test](https://github.com/ub-leipzig/api-constraint-validator) suite with Presentation API 2.1.1 manifests as the first test cases.  I am not sure that it is right for the general public, but for developers, it could be useful, since it can be run at the CLI for batch processing.  It is in Java 9 and uses JUnit 5.   The base [test methods](https://github.com/ub-leipzig/api-constraint-validator/blob/master/validator/src/test/java/de/ubleipzig/validator/RemoteManifestConstraintsTest.java) query or filter the test object as a graph, so it is fast and clean, and quite flexible.

The key task is to define the constraints from the verbose specification, and for this I made [a spreadsheet](https://github.com/ub-leipzig/api-constraint-validator/blob/master/docs/PresentationAPI%20Constraints.ods).  I plan to do this for v3 as well, and also for Web Annotations.

There are 29 test cases now and am following the JSON-LD standard test model (more or less).  There are certainly many more test cases to write, but it is a start.  Here is an [example report](https://htmlpreview.github.io/?https://github.com/ub-leipzig/api-constraint-validator/blob/master/summary-reports/test-summary_161017_1559.html).
